### ISSUE TITLE
Subtyping for narrowed types not always correct

### ISSUE THREAD

**Issue by kmillikin** [microsoft/pyright#4095]
In this very contrived example:

```python
class A: pass

class B: pass

class C: pass

def f(x: A):
  if isinstance(x, B):
    if not isinstance(x, C):
      return
  else:
    if not isinstance(x, C):
      return
  reveal_type(x)
```

Pyright has `<subclass of <subclass of A and B> and C> | <subclass of A and C>`.  The rules for unions have that we should eliminate an element of the union if it is a subtype of another.  `<subclass of <subclass of A and B> and C>` is a subtype of `<subclass of A and C>` and so the revealed type should be only `<subclass of A and C>`.
